### PR TITLE
INT-4542: CorrMH: Use popSequenceDetails properly

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-4542

Perform a `MessageBuilder.popSequenceDetails()` for those
`MessageGroupProcessor` results which are not `Message`
or `Collection<Message>`

### CODE REVIEW

File: spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java [hunk 1]
```diff
@@ -653,7 +654,8 @@ protected void forceComplete(MessageGroup group) {
 						 * groups. A longer timeout for empty groups can be enabled by
 						 * setting minimumTimeoutForEmptyGroups.
 						 */
-						removeGroup = lastModifiedNow <= (System.currentTimeMillis() - this.minimumTimeoutForEmptyGroups);
+						removeGroup = lastModifiedNow <= (System
+								.currentTimeMillis() - this.minimumTimeoutForEmptyGroups);
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java]: This looks odd; move System to the next line too?

File: spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java [hunk 2]
```diff
@@ -753,10 +755,24 @@ protected void completeGroup(Object correlationKey, MessageGroup group) {
...
+			if (result instanceof AbstractIntegrationMessageBuilder<?>) {
+				messageBuilder = (AbstractIntegrationMessageBuilder<?>) result;
+			}
+			else {
+				messageBuilder = getMessageBuilderFactory()
+						.withPayload(result)
+						.copyHeaders(message.getHeaders());
+			}
+			result = messageBuilder.popSequenceDetails();
+		}
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java]: Now I look at this, I wonder if it is completely valid to do this unconditionally. Consider

    splitter -> custom-splitter -> custom-aggregator -> aggregator

Perhaps we need a new flag `alwaysPopSequenceDetails` (true in 5.1, false in 5.0).

Or, add some metadata (`seqInfoId`)  to the sequence details and only pop if the splitter and aggregator are configured with the same metadata.

If the custom splitter doesn't push sequence details, we will remove the real ones from the first splitter.

**Review by garyrussell — CHANGES_REQUESTED** [review]

File: spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java (hunk 1 above)
Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java]: Hm. Some fault of my IDEA code reformatting.
Will fix.

File: spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java (hunk 2 above)
Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java]: OK. Sounds like we just need to be sure that our aggregation is based on the `sequenceDetails`. Because custom `splitter` may not be based on the `applySequence`.
So, maybe just an extra `sequenceApplied` header? But then how to understand on which level it has not been used?..

Or... I just don't see the whole picture and need to sleep with this...

Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java]: After sleeping with this, I came with the conclusion that your suggestion about `popSequenceDetails(true|false)` is the way to go.
We won't pollute messages with some extra header which may be really out of control and I don't want to tie an aggregator with the splitter.

We may do a condition based on the default `HeaderAttributeCorrelationStrategy(IntegrationMessageHeaderAccessor.CORRELATION_ID)` and `SimpleSequenceSizeReleaseStrategy`, but that still doesn't come as a robust solution since we can use defaults, but populate sequence header some other way, not via splitter.

From here I think we just need to add such a `popSequenceDetails` option to the correlation handler, make it `true` by default (looks like most cases, since we haven't had issues so far), modify a `AbstractAggregatingMessageGroupProcessor` to return a `AbstractIntegrationMessageBuilder` instead of the whole message and do conditional `.popSequenceDetails()` already in the `AbstractCorrelatingMessageHandler`. 

Saying that I don't think that we are good for back-porting to `5.0.x`: looks like this is going to be some breaking change.

Let me know WDYT?
Thanks

Comment (garyrussell) [reply spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java]: Maybe call it simply `popSequence` with the documentation explaining it undoes the nearest upstream `applySequence`?

Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java]: Makes sense. Will fix soon.

File: spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java [hunk 3]
```diff
@@ -217,6 +221,69 @@ public void setOutputProcessor(MessageGroupProcessor outputProcessor) {
...
+	 */
+	public void setExpireGroupsUponTimeout(boolean expireGroupsUponTimeout) {
+		this.expireGroupsUponTimeout = expireGroupsUponTimeout;
+	}
+
+	/**
+	 * Perform a {@link MessageBuilder#popSequenceDetails()} for output message or not.
+	 * Default to true.
+	 * This option plays an opposite role to the
+	 * {@link org.springframework.integration.splitter.AbstractMessageSplitter#setApplySequence(boolean)}.
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java]: It's not just splitter - router, pubsub etc; maybe just say "removes the sequence information added by the nearest upstream component with `applySequence=true` (for example splitter).

File: spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java (hunk 1 above)
Comment (garyrussell) [reply spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java]: Still weird indentation.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java [hunk 4]
```diff
@@ -531,6 +532,25 @@ public void testExceptionTypeRouteFlow() {
...
+		QueueChannel replyChannel = new QueueChannel();
+		Message<String> request = MessageBuilder.withPayload("this is a test")
+				.setReplyChannel(replyChannel)
+				.build();
+		this.nestedScatterGatherFlowInput.send(request);
+		Message<?> bestQuoteMessage = replyChannel.receive(10000);
+		assertNotNull(bestQuoteMessage);
+		Object payload = bestQuoteMessage.getPayload();
+		assertThat(payload, instanceOf(String.class));
+//		assertThat(((List<?>) payload).size(), greaterThanOrEqualTo(1));
```
Comment (garyrussell) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: Why is this here? The payload is a String.

File: src/reference/asciidoc/aggregator.adoc [hunk 5]
```diff
@@ -103,6 +103,13 @@ This method is invoked for aggregating messages as follows:
 
 NOTE: In the interest of code simplicity and promoting best practices such as low coupling, testability, and others, the preferred way of implementing the aggregation logic is through a POJO and using the XML or annotation support for configuring it in the application.
 
+Starting with version 5.1, after processing message group, an `AbstractCorrelatingMessageHandler` performs a `MessageBuilder.popSequenceDetails()` message headers modification for the proper splitter-aggregator scenario with several nested levels.
+It is done only if the message group release result is not a message or collection of messages.
+In that case a target `MessageGroupProcessor` is responsible for the `MessageBuilder.popSequenceDetails()` call during those messages building.
```
Comment (garyrussell) [inline src/reference/asciidoc/aggregator.adoc]: `s/during those messages building/while building those messages/`

File: src/reference/asciidoc/aggregator.adoc [hunk 6]
```diff
@@ -103,6 +103,13 @@ This method is invoked for aggregating messages as follows:
...
+This functionality can be controlled by newly introduced `popSequence` `boolean` property, so the `MessageBuilder.popSequenceDetails()` can be disabled in some scenarios when correlation details have not been populated by the standard splitter.
```
Comment (garyrussell) [inline src/reference/asciidoc/aggregator.adoc]: `s/by newly introduced/by a new/`

File: src/reference/asciidoc/aggregator.adoc [hunk 7]
```diff
@@ -103,6 +103,13 @@ This method is invoked for aggregating messages as follows:
...
+This property, essentially, undoes what has been done with the nearest upstream `applySequence = true` in the `AbstractMessageSplitter`.
```
Comment (garyrussell) [inline src/reference/asciidoc/aggregator.adoc]: ...by the nearest...

**Review by garyrussell — CHANGES_REQUESTED** [review]
Also needs a migration guide indicating that the sequence is now popped by default for all MGPs.

**Review by garyrussell — APPROVED** [review]

**Comment by garyrussell** [comment]
We're no longer cherry-picking; correct?

**Comment by artembilan** [comment]
> We're no longer cherry-picking; correct?

No, we are not. There is a workaround with the custom `MessageGroupProcessor` and explicit `popSequenceDetails()` call there.
Plus an `AbstractAggregatingMessageGroupProcessor` has a breaking change.

Migration Guide note has been added: https://github.com/spring-projects/spring-integration/wiki/Spring-Integration-5.0-to-5.1-Migration-Guide#aggregator-changes
